### PR TITLE
Streamline TestResult due to expectedExceptions

### LINKED ISSUE #2788: TestResult.isSuccess() is TRUE when test fails due to expectedExceptions
### TestNG Version
7.6.1

### Expected behavior
`IInvokedMethodListener.afterInvocation(method, result)` should receive a result with the status `FAILURE`.

### Actual behavior
The status is `SUCCESS` and `getThrowable()` is set to a `TestException` from [ExpectedExceptionsHolder.noException](https://github.com/cbeust/testng/blob/4de3c7b08a5ffd77cdc23277fda45b9dadca949d/testng-core/src/main/java/org/testng/internal/invokers/ExpectedExceptionsHolder.java#L73-L82).

### Is the issue reproducible on runner?
Yes, minimally

- [x] Gradle
- [x] Eclipse

### Test case sample
```java
@Test(expectedExceptions = NullPointerException.class)
public void test() {}
```

### Impact

To handle running 10M+ unit tests, my `IInvokedMethodListener` tries to reclaim memory for successful tests. This avoids out-of-memory errors and long GC pauses, e.g. by replacing test parameters with their stringified version. This also unset the throwable, which causes Gradle's runner to fail with an obscure NPE (https://github.com/gradle/gradle/issues/21336) as it was not defensive to failures without a cause (whereas the Eclipse plugin is tolerant). My fix is to stop nulling out the throwable as it is lightweight enough to retain. However, others might similarly be surprised by the premature success and report it incorrectly in their logic.

### Possible Fix

I suspect that adding `testResult.setStatus(ITestResult.FAILURE)` to the below code would be the correct change. That requires a little rework to retain the `originalStatus`.

https://github.com/cbeust/testng/blob/b5b3e1d0f8db429b1bb5d1d675ddf5940d19f92c/testng-core/src/main/java/org/testng/internal/invokers/TestInvoker.java#L865-L872

### PR DESCRIPTION
**Pull request by krmahadevan**
Closes #2788

Fixes #2788 .

### Did you remember to?

- [X] Add test case(s)
- [X] Update `CHANGES.txt`
- [X] Auto applied styling via `./gradlew autostyleApply`

We encourage pull requests that:

* Add new features to TestNG (or)
* Fix bugs in TestNG

If your pull request involves fixing SonarQube issues then we would suggest that you please discuss this with the 
[TestNG-dev](https://groups.google.com/forum/#!forum/testng-dev) before you spend time working on it.

**Note:** For more information on contribution guidelines  please make sure you refer our [Contributing](.github/CONTRIBUTING.md) section for detailed set of steps.

### CODE REVIEW

File: testng-core/src/main/java/org/testng/internal/invokers/TestInvoker.java [hunk 1]
```diff
@@ -726,6 +726,9 @@ private ITestResult invokeMethod(
       StatusHolder holder =
           considerExceptions(
               arguments.getTestMethod(), testResult, expectedExceptionClasses, failureContext);
+      if (holder.status != testResult.getStatus()) {
```
Comment (juherr) [inline testng-core/src/main/java/org/testng/internal/invokers/TestInvoker.java:729]: No need to check, you can always set the value. Please add a comment that explains **why** we override the status here.

File: testng-core/src/main/java/org/testng/internal/ExitCode.java [hunk 2]
```diff
@@ -19,7 +19,7 @@ public class ExitCode {
   public static final int HAS_NO_TEST = 8;
   private static final int FAILED_WITHIN_SUCCESS = 4;
   public static final int SKIPPED = 2;
-  private static final int FAILED = 1;
+  public static final int FAILED = 1;
```
Comment (juherr) [inline testng-core/src/main/java/org/testng/internal/ExitCode.java:22]: Add comment: visible for tests

Comment (krmahadevan) [reply testng-core/src/main/java/org/testng/internal/ExitCode.java:22]: This is a constant. Its not going to be altered. The comment is just going to be redundant IMO

Comment (juherr) [reply testng-core/src/main/java/org/testng/internal/ExitCode.java:22]: Ok

File: testng-core/src/main/java/org/testng/internal/invokers/TestInvoker.java (hunk 1 above)
Comment (krmahadevan) [reply testng-core/src/main/java/org/testng/internal/invokers/TestInvoker.java]: Fixed

**Comment by krmahadevan** [comment]
@juherr - I amended the commit and added the suggested changes. Please take a look

**Review by juherr — APPROVED** [review]

**Pull request merged**